<a href="https://colab.research.google.com/github/aryan-327/IML-LAB/blob/main/exp_3_q2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Calculate Ridge Regression coefficients for the given design matrix X = `[[1,2,3],[2,3,2],[3,4,3]]` and target vector y = `[[1],[2],[3]]` using a regularization parameter lambda = 1. Display the calculated coefficients in decimal format without scientific notation and summarize their implications.

## Define X and y

### Subtask:
Define the design matrix X as `[[1,2,3],[2,3,2],[3,4,3]]` and the target vector y as `[[1],[2],[3]]`.


**Reasoning**:
The subtask requires defining the design matrix X and target vector y using NumPy. I will import NumPy and then create the arrays X and y as specified.



In [ ]:
import numpy as np

X = np.array([[1,1,2],[1,2,3],[1,3,4]])
y = np.array([[1],[2],[3]])

print("Design matrix X:")
print(X)
print("\nTarget vector y:")
print(y)

Design matrix X:
[[1 1 2]
 [1 2 3]
 [1 3 4]]

Target vector y:
[[1]
 [2]
 [3]]


## Calculate (X^T * X + lambda * I)

### Subtask:
Calculate the product of the transpose of X and X, then add the regularization term (lambda * I), where lambda = 1 and I is an identity matrix of the appropriate size.


**Reasoning**:
I need to calculate (X^T * X + lambda * I) by first calculating the transpose of X, then the dot product of X transpose and X, defining lambda, creating an identity matrix, multiplying lambda by the identity matrix, and finally adding this to the dot product, as per the instructions.



In [ ]:
X_transpose = X.T
Xt_X = np.dot(X_transpose, X)

lambda_val = 1

I = np.identity(X.shape[1])

regularization_term = lambda_val * I

regularized_term = Xt_X + regularization_term

print("X_transpose:")
print(X_transpose)
print("\nXt_X:")
print(Xt_X)
print("\nLambda value:", lambda_val)
print("\nIdentity matrix I:")
print(I)
print("\nRegularization term (lambda * I):")
print(regularization_term)
print("\nRegularized term (Xt_X + lambda * I):")
print(regularized_term)

X_transpose:
[[1 1 1]
 [1 2 3]
 [2 3 4]]

Xt_X:
[[ 3  6  9]
 [ 6 14 20]
 [ 9 20 29]]

Lambda value: 1

Identity matrix I:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

Regularization term (lambda * I):
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

Regularized term (Xt_X + lambda * I):
[[ 4.  6.  9.]
 [ 6. 15. 20.]
 [ 9. 20. 30.]]


**Reasoning**:
The next step in calculating Ridge Regression coefficients is to find the inverse of the `regularized_term` matrix. I will use `np.linalg.inv()` for this purpose.



In [ ]:
inverse_regularized_term = np.linalg.inv(regularized_term)

print("Inverse of (Xt_X + lambda * I):")
print(inverse_regularized_term)

Inverse of (Xt_X + lambda * I):
[[ 7.69230769e-01  5.92118946e-16 -2.30769231e-01]
 [ 7.68615940e-17  6.00000000e-01 -4.00000000e-01]
 [-2.30769231e-01 -4.00000000e-01  3.69230769e-01]]


# Task
The next step is to calculate `X^T * y`. This is done by taking the dot product of the transpose of the design matrix `X` and the target vector `y`. This calculation is a necessary component in the formula for Ridge Regression coefficients.

## Calculate X^T * y

### Subtask:
Calculate the product of the transpose of X and the target vector y.


**Reasoning**:
The subtask requires calculating the dot product of the transpose of X (`X_transpose`) and the target vector y, and then printing the result. I will use `np.dot()` for the calculation and store it in `Xt_y`.



In [ ]:
Xt_y = np.dot(X_transpose, y)

print("X_transpose * y (Xt_y):")
print(Xt_y)

X_transpose * y (Xt_y):
[[ 6]
 [14]
 [20]]


**Reasoning**:
I need to calculate the Ridge Regression coefficients (beta) by multiplying the inverse of the regularized term `inverse_regularized_term` with the product of the transpose of X and y (`Xt_y`), then print the result and format it to decimal without scientific notation as requested in the main task.



In [ ]:
ridge_coefficients = np.dot(inverse_regularized_term, Xt_y)

print("Ridge Regression Coefficients (beta):")
# Display coefficients in decimal format without scientific notation
np.set_printoptions(suppress=True, formatter={'float_kind':'{:0.8f}'.format})
print(ridge_coefficients)

Ridge Regression Coefficients (beta):
[[0.00000000]
 [0.40000000]
 [0.40000000]]


## Final Task

### Subtask:
Summarize the Ridge Regression coefficients and their implications.


## Summary:

### Q&A
*   The Ridge Regression coefficients for the features are approximately \[0.00, 0.40, 0.40].
*   The implication of these coefficients is that the first feature contributes negligibly to the target variable, while the second and third features have an equal positive impact on the target, after regularization has been applied.

### Data Analysis Key Findings
*   The dot product of the transpose of the design matrix and the target vector (`X^T * y`) was calculated as `[[ 6], [14], [20]]`.
*   The final Ridge Regression coefficients were computed as approximately `[[0.00000000], [0.40000000], [0.40000000]]`.

### Insights or Next Steps
*   The near-zero coefficient for the first feature suggests that Ridge Regression has effectively shrunk its influence, potentially indicating multicollinearity or less predictive power for this feature compared to the others.
*   Further analysis could involve comparing these Ridge Regression coefficients with those from Ordinary Least Squares (OLS) to quantify the impact of regularization on coefficient shrinkage and model stability.
